sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/AppendBlobAsyncApiTests.java

```diff
@@ -0,0 +1,915 @@
+        Map<String, String> tags = new HashMap<>();
+        if (key1 != null) {
+            tags.put(key1, value1);
+        }
+        if (key2 != null) {
+            tags.put(key2, value2);
+        }
+
+        bc.createIfNotExistsWithResponse(new AppendBlobCreateOptions().setTags(tags)).block();
+
+        StepVerifier.create(bc.getTagsWithResponse(new BlobGetTagsOptions()))
+            .assertNext(r -> {
+                for (Map.Entry<String, String> entry : tags.entrySet()) {
+                    assertEquals(entry.getValue(), r.getValue().get(entry.getKey()));
+                }
+            })
+            .verifyComplete();
+    }
+
+    private Stream<Arguments> createIfNotExistsTagsSupplier() {
+        return Stream.of(
+            Arguments.of(null, null, null, null),
+            Arguments.of("foo", "bar", "fizz", "buzz"),
+            Arguments.of(" +-./:=_  +-./:=_", " +-./:=_", null, null));
+    }
+
+    @Test
+    public void appendBlockDefaults() {
+        StepVerifier.create(bc.appendBlockWithResponse(DATA.getDefaultFlux(),
+            DATA.getDefaultDataSize(), null, null))
+            .assertNext(r -> {
+                validateBasicHeaders(r.getHeaders());
+                assertNotNull(r.getHeaders().getValue(X_MS_CONTENT_CRC64));
+                assertNotNull(r.getValue().getBlobAppendOffset());
+                assertNotNull(r.getValue().getBlobCommittedBlockCount());
+            })
+            .verifyComplete();
+
+        StepVerifier.create(FluxUtil.collectBytesInByteBufferStream(bc.downloadStream()))
+            .assertNext(r -> TestUtils.assertArraysEqual(DATA.getDefaultBytes(), r))
+            .verifyComplete();
+    }
+
+    @Test
+    public void appendBlockMin() {
+        assertAsyncResponseStatusCode(bc.appendBlockWithResponse(DATA.getDefaultFlux(),
+            DATA.getDefaultDataSize(), null, null), 201);
+    }
+
+    @ParameterizedTest
+    @MethodSource("appendBlockIASupplier")
+    public void appendBlockIA(Flux<ByteBuffer> stream, long dataSize, Class<? extends Throwable> exceptionType) {
+        StepVerifier.create(bc.appendBlock(stream, dataSize))
+            .verifyError(exceptionType);
+    }
+
+    private static Stream<Arguments> appendBlockIASupplier() {
+        return Stream.of(
+            Arguments.of(null, DATA.getDefaultDataSize(), NullPointerException.class),
+            Arguments.of(DATA.getDefaultFlux(), DATA.getDefaultDataSize() + 1, UnexpectedLengthException.class),
+            Arguments.of(DATA.getDefaultFlux(), DATA.getDefaultDataSize() - 1, UnexpectedLengthException.class)
+        );
+    }
+
+    @Test
+    public void appendBlockEmptyBody() {
+        StepVerifier.create(bc.appendBlock(Flux.just(ByteBuffer.wrap(new byte[0])), 0))
+            .verifyError(BlobStorageException.class);
+    }
+
+    @Test
+    public void appendBlockNullBody() {
+        assertThrows(NullPointerException.class, () -> bc.appendBlock(Flux.just((ByteBuffer) null), 0));
```

Comment:
Initially this test was failing completely because the chain of async appendBlock methods never actually checked if the data was null - a similar problem that I've seen with previous tests. 

I added this line to appendBlockWithResponse: 

`if (data == null) {
     return Mono.error(new NullPointerException("'data' cannot be null."));
}`

Let me know if you think this is the correct way to do this.

Reply:
That's the correct way to do it 😄